### PR TITLE
[Parse] Avoid skipping bodies with `/.../` regex literals

### LINKED ISSUE #59493: Regex literal with `"` fails to compile in items with `Unterminated string literal`
```swift
let test = /test":"(.*?)"/

enum Regexes
  { static let test = /test":"(.*?)"/ }
```

The first compiles and works, the second fails to compile with `Unterminated string literal`.

Swift 5.7 with Xcode 14 Beta 1 on macOS 12.4.

### PR DESCRIPTION
**Pull request by hamishknight**
While skipping, if we encounter a token that looks like it could be the start of a `/.../` regex literal, fall back to parsing the function or type body normally, as such a token could become a regex literal. As such, it could treat `{` and `}` as literal, or otherwise have contents that would be lexically invalid Swift.

To avoid falling back in too many cases, we apply the existing regex literal heuristics. Cases that pass the heuristic fall back to regular parsing. Cases that fail the heuristic are further checked to make sure they wouldn't contain an unbalanced `{` or `}`, but otherwise are allowed to be skipped. This allows us to continue skipping for most occurrences of infix and prefix `/`.

This is meant as a lower risk workaround to fix the the issue, we ought to go back to handling regex literals in the lexer.

Resolves https://github.com/apple/swift/issues/59493
Resolves rdar://95354010

### CODE REVIEW

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test macOS

**Comment by ahoppen** [comment]
@swift-ci Please SourceKit stress test

File: lib/Parse/Lexer.cpp [hunk 1]
```diff
@@ -1978,27 +1980,76 @@ const char *Lexer::findEndOfCurlyQuoteStringLiteral(const char *Body,
...
 }
 
-bool Lexer::tryLexRegexLiteral(const char *TokStart) {
+bool Lexer::isPotentialUnskippableBareSlashRegexLiteral(const Token &Tok) const {
+  if (!LangOpts.hasFeature(Feature::BareSlashRegexLiterals))
+    return false;
+
+  // A `/.../` regex literal may only start on a binary or prefix operator.
+  if (Tok.isNot(tok::oper_prefix, tok::oper_binary_spaced,
+                tok::oper_binary_unspaced)) {
```
[the remark this answers is not in the document]
Comment (hamishknight) [reply lib/Parse/Lexer.cpp:1989]: I can add some

**Comment by hamishknight** [comment]
@swift-ci please test

**Review by rintaro — APPROVED** [review]

**Comment by hamishknight** [comment]
@swift-ci please test macOS

**Comment by hamishknight** [comment]
@swift-ci please test macOS

**Pull request merged**